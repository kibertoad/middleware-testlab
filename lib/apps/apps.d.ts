export interface EndpointDefinition {
  method: 'get' | 'post' | 'delete' | 'put' | 'patch' | 'options' | 'head'
  path: string
}
