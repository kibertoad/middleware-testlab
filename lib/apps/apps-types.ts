import { Request, Response } from 'express-serve-static-core'
import { BaseContext } from 'koa'

export interface EndpointDefinition {
  method: 'get' | 'post' | 'delete' | 'put' | 'patch' | 'options' | 'head'
  path: string
}

export interface ExpressEndpointAssertor {
  (request: Request): void
}

export interface ExpressEndpointResponseAssertor {
  (response: Response): void
}

export interface ErrorAssertor {
  (error: any): void
}

export interface KoaEndpointAssertor {
  (ctx: BaseContext): void
}
