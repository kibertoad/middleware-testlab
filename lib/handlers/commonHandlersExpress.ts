import { Request, Response } from 'express'

export const okHandler = (_req: Request, res: Response) => {
  res.status(200).json({
    status: 'OK'
  })
}
