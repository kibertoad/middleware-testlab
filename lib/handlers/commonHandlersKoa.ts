import { BaseContext } from 'koa'

export const okHandler = (ctx: BaseContext, next: Function) => {
  ctx.status = 200
  ctx.body = {
    status: 'OK'
  }
  next()
}
