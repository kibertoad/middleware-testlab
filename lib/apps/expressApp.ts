import { Application, Request, Response } from 'express'
import express, { Router } from 'express'
import { RequestHandlerParams } from 'express-serve-static-core'
import { EndpointDefinition } from './apps'

export const DEFAULT_ENDPOINT = '/'
const DEFAULT_HANDLER = (_req: Request, res: Response, next: Function) => {
  try {
    res.status(201).send
  } catch (e) {
    console.error('Error while processing request: ', e)
    next(e)
  }
}

/**
 *
 * @param {RequestHandlerParams[]} middleware
 * @param {RequestHandlerParams} [endpointHandler] - if handler is not specified, by default sends 204 NO CONTENT
 * @param {string | EndpointDefinition} [endpoint='/'] - if endpoint is passed as string, or not specified, by default GET method is used.
 */
export function newExpressApp(
  middleware: RequestHandlerParams[],
  endpointHandler: RequestHandlerParams = DEFAULT_HANDLER,
  endpoint: string | EndpointDefinition = DEFAULT_ENDPOINT
): Application {
  const app = express()
  middleware.forEach((middlewareEntry: RequestHandlerParams) => {
    app.use(middlewareEntry)
  })

  // Resolve endpoint from params
  let method: string
  let path: string
  if (typeof endpoint === 'string') {
    method = 'get'
    path = endpoint
  } else {
    method = endpoint.method
    path = endpoint.path
  }

  const router = Router()
  // @ts-ignore
  router[method](path, endpointHandler)
  app.use(router)
  return app
}
