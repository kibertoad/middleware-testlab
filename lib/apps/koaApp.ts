import Koa, { BaseContext, Middleware } from 'koa'
import Router from 'koa-router'
import Application from 'koa'
import { EndpointDefinition } from './apps'

export const DEFAULT_ENDPOINT = '/'
const DEFAULT_HANDLER = (ctx: BaseContext, next: Function) => {
  ctx.status = 204
  next()
}

/**
 *
 * @param {Middleware[]} [appMiddleware=[]]
 * @param {Middleware[]} [routerMiddleware=[]]
 * @param {Middleware} [handler] - if handler is not specified, by default sends 204 NO CONTENT
 * @param {string | EndpointDefinition} [endpoint='/'] - if endpoint is passed as string, or not specified, by default GET method is used.
 */
export function newKoaApp({
  appMiddleware = [],
  routerMiddleware = [],
  handler = DEFAULT_HANDLER,
  endpoint = DEFAULT_ENDPOINT
}: {
  appMiddleware?: Middleware[]
  routerMiddleware?: Middleware[]
  handler?: Middleware
  endpoint?: string | EndpointDefinition
}): Application {
  const app = new Koa()
  appMiddleware.forEach((middlewareEntry: Middleware) => {
    app.use(middlewareEntry)
  })

  const router = new Router()

  routerMiddleware.forEach((middlewareEntry: Middleware) => {
    router.use(middlewareEntry)
  })

  // Resolve endpoint from params
  let method: string
  let path: string
  if (typeof endpoint === 'string') {
    method = 'get'
    path = endpoint
  } else {
    method = endpoint.method
    path = endpoint.path
  }

  // @ts-ignore
  router[method](path, handler)
  app.use(router.routes()).use(router.allowedMethods())

  return app
}
