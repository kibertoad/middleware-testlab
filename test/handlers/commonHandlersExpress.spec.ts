import { DEFAULT_ENDPOINT, newExpressApp } from '../../lib/apps/expressApp'
import { throwingHandler, okHandler } from '../../lib/handlers/commonHandlersExpress'
import request from 'supertest'

describe('commonHandler', () => {
  describe('okHandler', () => {
    it('happy path', async () => {
      const app = newExpressApp({
        handler: okHandler
      })

      const response = await request(app).get(DEFAULT_ENDPOINT)

      expect(response.body).toEqual({
        status: 'OK'
      })
      expect(response.status).toEqual(200)
    })
  })

  describe('throwingHandler', () => {
    it('happy path', async () => {
      const app = newExpressApp({
        handler: throwingHandler
      })

      const response = await request(app).get(DEFAULT_ENDPOINT)

      expect(response.body).toMatchObject({
        message: 'Something broke down',
        name: 'Error',
        details: {
          timestamp: '2019-11-6'
        }
      })
      expect(response.status).toEqual(500)
    })
  })
})
