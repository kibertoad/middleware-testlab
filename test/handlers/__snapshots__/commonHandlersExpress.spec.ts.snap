// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commonHandler okHandler error assertors throw an error when no error is raised 1`] = `
Object {
  "message": "Expected error to be thrown, but it wasn't",
  "name": "Error",
}
`;
